lib/node.js

```diff
@@ -81,36 +75,181 @@ Node.prototype._loadBitcoind = function(config) {
+      }
+
+      done(null, ancestorHash);
+    });
+  });
+};
+
+/**
+ * This function will attempt to rewind the chain to the common ancestor
+ * between the current chain and a forked block.
+ * @param {Block} block - The new tip that forks the current chain.
+ * @param {Function} done - A callback function that is called when complete.
+ */
+Node.prototype._syncBitcoindRewind = function(block, done) {
+
+  var self = this;
+
+  self._syncBitcoindAncestor(block, function(err, ancestorHash) {
+
+    // Rewind the chain to the common ancestor
+    async.whilst(
+      function() {
+        // Wait until the tip equals the ancestor hash
+        return self.chain.tip.hash !== ancestorHash;
+      },
+      function(removeDone) {
+
+        var tip = self.chain.tip;
+
+        self.getBlock(tip.prevHash, function(err, previousTip) {
+          if (err) {
+            removeDone(err);
+          }
+
+          // Undo the related indexes for this block
+          self.db._onChainRemoveBlock(tip, function(err) {
+            if (err) {
+              return removeDone(err);
+            }
+
+            // Set the new tip
+            delete self.chain.tip.__transactions;
+            previousTip.__height = self.chain.tip.__height - 1;
+            self.chain.tip = previousTip;
+            self.chain.saveMetadata();
+            self.chain.emit('removeblock', tip);
+            removeDone();
+          });
+
+        });
+
+      }, done
+    );
+  });
+};
+
+/**
+ * This function will synchronize additional indexes for the chain based on
+ * the current active chain in the bitcoin daemon. In the event that there is
+ * a reorganization in the daemon, the chain will rewind to the last common
+ * ancestor and then resume syncing.
+ */
 Node.prototype._syncBitcoind = function() {
   var self = this;
 
+  if (self.bitcoindSyncing) {
+    return;
+  }
+
+  if (!self.chain.tip) {
+    return;
+  }
+
+  self.bitcoindSyncing = true;
+
   log.info('Starting Bitcoind Sync');
 
-  var info = self.bitcoind.getInfo();
   var height;
 
   async.whilst(function() {
-    if (self.syncStrategy !== Node.SYNC_STRATEGIES.BITCOIND) {
-      log.info('Stopping Bitcoind Sync');
-      return false;
-    }
     height = self.chain.tip.__height;
-    return height < info.blocks;
-  }, function(next) {
+    return height < self.bitcoindHeight;
+  }, function(done) {
     self.bitcoind.getBlock(height + 1, function(err, blockBuffer) {
       if (err) {
-        return next(err);
+        return done(err);
+      }
+      var block = self.Block.fromBuffer(blockBuffer);
+      if (block.prevHash === self.chain.tip.hash) {
+
+        // This block appends to the current chain tip and we can
+        // immediately add it to the chain and create indexes.
+
+        // Populate height
+        block.__height = self.chain.tip.__height + 1;
+
+        // Create indexes
+        self.db._onChainAddBlock(block, function(err) {
```

Comment:
You're missing some stuff that chain.addBlock() does when it adds a block - for instance updating the hashes cache. Is there a reason you are calling db._onChainAddBlock directly? I'm definitely not opposed to making chain really lean like it is in bitcore-node and then putting most of the logic in Node, like you're starting to do here. But we should make sure we move all the functionality out of chain and into node if we do that. And that'll be a big refactor touching our other projects.